trieste/bayesian_optimizer.py

```diff
@@ -444,25 +445,33 @@ def optimize(
         models = cast(
             Dict[str, TrainableProbabilisticModelType], map_values(create_model, model_specs)
         )
+
         history: list[Record[StateType]] = []
 
         for step in range(num_steps):
             set_step_number(step)
             try:
-                if track_state:
-                    models_copy = copy.deepcopy(models)
-                    acquisition_state_copy = copy.deepcopy(acquisition_state)
-                    history.append(Record(datasets, models_copy, acquisition_state_copy))
+
+                t_0_total = time()  # start wallclock for total BO step time
 
                 if step == 0 and fit_initial_model:
+                    t_0_model = time()  # start wallclock for initial model fitting
```

Comment:
You can make these timing calculations nicer (and less error prone) by defining a timing context manager in trieste.utils:

```python
from time import perf_counter
from contextlib import contextmanager

@contextmanager
def timeit() -> float:
    t_0 = perf_counter()
    yield lambda: perf_counter() - t_0
```

which will let you simplify the code to:

```python
with timeit() as initial_model_fitting_time:
    for tag, model in models.items():
        ...
```

You can nest timeits so the total timer will work too.

Reply:
Much nicer!